File: Backtracking/679. 24 Game/solution.py
class Solution:
    def judgePoint24(self, cards: List[int]) -> bool:
        permutations = []
        def getPerm(state, bitmask):
            if len(state) == 4:
                permutations.append(state.copy())
                return
            for i in range(4):
                if (1<<i)&bitmask: continue
                bitmask |= (1<<i)
                getPerm(state + [cards[i]], bitmask)
                bitmask ^= (1<<i)
        getPerm([], 0) # or itertools.permutations(cards)

        def dfs(nums, l, r):
            if l == r: return {nums[l]}

            res = set()
            for i in range(l, r):
                left = dfs(nums, l, i)
                right = dfs(nums, i+1, r)
                for x in left:
                    for y in right:
                        res.add(x+y)
                        res.add(x-y)
                        res.add(x*y)
                        if y != 0:
                            res.add(x/y)
            return res

        PRECISION = 1e-10
        for nums in permutations:
            SET = dfs(nums, 0, 3)
            for res in SET:
                if abs(res-24) < PRECISION: return True
        return False

File: Backtracking/679. 24 Game/solution.cpp
class Solution {
public:
    bool judgePoint24(vector<int>& cards) {
        sort(cards.begin(), cards.end());
                
        double epsilon = 1e-6;
        do 
        {
            unordered_set<double> results = dfs(cards, 0, 3);
            for (auto res:results) {
                if (abs(res-24) < epsilon) return true;
            }
        } while (next_permutation(cards.begin(), cards.end()));

        return false;
    }
    
    unordered_set<double> dfs(vector<int> &nums, int l, int r) {
        if (l==r) return {(double) nums[l]};
        
        unordered_set<double>res;
        
        for (int i=l; i<r; i++) {
            unordered_set<double>left = dfs(nums,l,i);
            unordered_set<double>right = dfs(nums,i+1,r);
            for (double x:left)
                for (double y:right) {
                    res.insert(x+y);
                    res.insert(x-y);
                    res.insert(x*y);
                    if (y!=0) {
                        res.insert(x/y);
                    }
                }
        }
        return res;
    }
};